File: prep/implementation/patterns_reference_implementation/merge_intervals/insert_interval.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

vector<Interval> insert(vector<Interval> &a, Interval &inter) {
  vector<Interval> res;

  int i = 0;
  while (i < a.size() && a[i].end < inter.start) {
    res.emplace_back(Interval(a[i].start, a[i].end));
    i++;
  }

  while (i < a.size() && a[i].start <= inter.end) {
    inter.start = min(inter.start, a[i].start);
    inter.end = max(inter.end, a[i].end);
    i++;
  }

  res.emplace_back(inter);

  while (i < a.size()) {
    res.emplace_back(a[i++]);
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    int s, e;

    for (int i = 0; i < n; i++) {
      cin >> s >> e;
      a.emplace_back(Interval(s, e));
    }

    cin >> s >> e;
    Interval inter(s, e);

    auto res = insert(a, inter);
    for (auto e : res) {
      cout << e.start << " " << e.end << "\n";
    }
  }

  return 0;
}

File: prep/implementation/patterns_reference_implementation/merge_intervals/InsertInterval.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class InsertInterval {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int s = in.nextInt();
          int e = in.nextInt();

          a.add(new Interval(s, e));
        }

        int s = in.nextInt();
        int e = in.nextInt();
        Interval newInterval = new Interval(s, e);

        List<Interval> res = new ArrayList<>();

        int i = 0;
        while (i < n && a.get(i).end < newInterval.start) {
          res.add(a.get(i));
          i++;
        }

        while (i < n && a.get(i).start <= newInterval.end) {
          newInterval.start = Math.min(newInterval.start, a.get(i).start);
          newInterval.end = Math.max(newInterval.end, a.get(i).end);
          i++;
        }

        res.add(newInterval);

        while (i < n) {
          res.add(a.get(i));
          i++;
        }

        for (var inter : res) {
          System.out.printf("%d %d\n", inter.start, inter.end);
        }
      }
    }
  }
}